Signal/src/ViewControllers/SendExternalFileViewController.m

```diff
@@ -0,0 +1,146 @@
+
+    _messageSender = [Environment getCurrent].messageSender;
+
+    self.title = NSLocalizedString(@"SEND_EXTERNAL_FILE_VIEW_TITLE", @"Title for the 'send external file' view.");
+}
+
+#pragma mark - SelectThreadViewControllerDelegate
+
+- (void)threadWasSelected:(TSThread *)thread
+{
+    OWSAssert(self.attachment);
+    OWSAssert(thread);
+
+    [ThreadUtil sendMessageWithAttachment:self.attachment inThread:thread messageSender:self.messageSender];
+
+    [Environment messageThreadId:thread.uniqueId];
+}
+
+- (BOOL)canSelectBlockedContact
+{
+    return NO;
+}
+
+- (nullable UIView *)createHeader:(UIView *)superview
+{
+    OWSAssert(superview)
+
+        const CGFloat imageSize
+        = ScaleFromIPhone5To7Plus(40, 40);
+    const CGFloat imageLabelSpacing = ScaleFromIPhone5To7Plus(5, 5);
+    const CGFloat titleVSpacing = ScaleFromIPhone5To7Plus(10, 10);
+    const CGFloat contentVMargin = ScaleFromIPhone5To7Plus(20, 20);
+
+    UIView *header = [UIView new];
+    [superview addSubview:header];
+    [header autoPinWidthToSuperview];
+    [header autoPinToTopLayoutGuideOfViewController:self withInset:0];
+
+    UIView *titleLabel = [self createTitleLabel];
+    [header addSubview:titleLabel];
+    [titleLabel autoHCenterInSuperview];
+    [titleLabel autoPinEdgeToSuperviewEdge:ALEdgeTop withInset:contentVMargin];
+
+    UIView *fileView = [UIView new];
+    [header addSubview:fileView];
+    [fileView autoHCenterInSuperview];
+    [fileView autoPinEdgeToSuperviewEdge:ALEdgeBottom withInset:contentVMargin];
+    [fileView autoPinEdge:ALEdgeTop toEdge:ALEdgeBottom ofView:titleLabel withOffset:titleVSpacing];
+
+    UIImage *image = [UIImage imageNamed:@"file-thin-black-large"];
+    OWSAssert(image);
+    UIImageView *imageView = [[UIImageView alloc] initWithImage:image];
+    imageView.layer.minificationFilter = kCAFilterTrilinear;
+    imageView.layer.magnificationFilter = kCAFilterTrilinear;
+    [fileView addSubview:imageView];
+    [imageView autoSetDimension:ALDimensionWidth toSize:imageSize];
+    [imageView autoSetDimension:ALDimensionHeight toSize:imageSize];
+    [imageView autoPinEdgeToSuperviewEdge:ALEdgeLeft];
+    [imageView autoPinEdgeToSuperviewEdge:ALEdgeTop];
+    [imageView autoPinEdgeToSuperviewEdge:ALEdgeBottom];
+
+    UIView *fileNameLabel = [self createFileNameLabel];
+    [fileView addSubview:fileNameLabel];
+    [fileNameLabel autoAlignAxis:ALAxisHorizontal toSameAxisOfView:imageView];
+    [fileNameLabel autoPinEdge:ALEdgeLeft toEdge:ALEdgeRight ofView:imageView withOffset:imageLabelSpacing];
+    [fileNameLabel autoPinEdgeToSuperviewEdge:ALEdgeRight];
+
+    return header;
+}
+
+- (NSString *)formattedFileName
+{
+    OWSAssert(self.attachment) NSString *filename =
+        [self.attachment.filename stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
+    OWSAssert(filename.length > 0);
+    const NSUInteger kMaxFilenameLength = 50;
+    if (filename.length > kMaxFilenameLength) {
```

Comment:
I'm taking out `NSLineBreakByTruncatingMiddle`.  It's because we want to center the filename & the file icon.  We could do this using `NSLineBreakByTruncatingMiddle` and a low-priority constraint on the width of the wrapping view, but I try to avoid using priority in my auto-layout code as much as possible because it _very_ quickly leads to unmaintainable code.